Реализовал систему хештегов с REST API и тестами

vzeten opened the pull request:
Реализовал систему хештегов для постов

Функциональность:
- Добавлена сущность Hashtag со ссылкой на post_id
- Создан HashtagRepository с кастомными запросами
- Реализованы интерфейс HashtagService и HashtagServiceImpl:
    * Извлечение хештегов из контента поста с помощью regex
    * Сохранение хештегов в базу данных
    * Поиск постов по названию хештега с пагинацией
- Добавлен REST-контроллер HashtagController:
    * GET /api/hashtags/{name}/posts с поддержкой пагинации
    * Валидация входных данных (page >= 0, size 1-100)
- Добавлен PostDto record для API-ответов
- Настроен PostMapper с помощью MapStruct

Тесты:
- Добавлено 14 юнит-тестов для HashtagServiceImpl
- Покрыты все методы: extractHashtagNames, saveHashtags, findPostsByHashtagName
- Протестированы граничные случаи: пустой контент, пагинация, несуществующие хештеги

База данных:
- Добавлена миграция V005__hashtag.sql с таблицей hashtag
- Создан индекс на имени хештега для повышения производительности

Стек технологий: Spring Boot, Spring Data JPA, MapStruct, Mockito, JUnit 5

Commit:
Реализовал систему хештегов с REST API и тестами

- Добавлена сущность Hashtag и репозиторий
- Реализован HashtagService с извлечением хештегов и поиском
- Добавлен REST-эндпоинт для поиска постов по хештегу
- Добавлены PostDto record и PostMapper
- Написаны юнит-тесты с полным покрытием HashtagServiceImpl
- Добавлена миграция базы данных для таблицы хештегов

AlexeyAntipin commented on src/main/java/faang/school/postservice/service/HashtagService.java:
```diff
@@ -0,0 +1,19 @@
+
+public interface HashtagService {
+
+    // Извлечь имена хештегов из текста
```
Давай все-таки java-doc)
А то некрасиво как-то

AlexeyAntipin commented on src/main/java/faang/school/postservice/service/HashtagServiceImpl.java:
```diff
@@ -0,0 +1,112 @@
+        log.info("Finding posts by hashtag: '{}', page: {}, size: {}",
+                name, pageable.getPageNumber(), pageable.getPageSize());
+
+        // 1. Найти все ID постов с этим хештегом
```
Кажется, что комменты бессмысленны, и так код вполне понятный

AlexeyAntipin commented on src/main/java/faang/school/postservice/service/HashtagServiceImpl.java:
```diff
@@ -0,0 +1,112 @@
+        // Проверка: если нет постов с таким хештегом
+        if (postIds.isEmpty()) {
+            log.info("No posts found for hashtag: '{}'", name);
+            return new PageImpl<>(List.of(), pageable, 0);
```
Лучше вместо List.of() использовать Collections.emptyList()

AlexeyAntipin commented on src/main/java/faang/school/postservice/repository/HashtagRepository.java:
```diff
@@ -0,0 +1,18 @@
+
+    List<Hashtag> findAllByName(String name);
+
+    @Query("SELECT DISTINCT h.postId FROM Hashtag h WHERE h.name = :name")
```
Можно этот метод реализовать без query. jpa поддерживает подобные запросы чисто по названию метода

Вот тут доку можно почитать: https://docs.spring.io/spring-data/jpa/reference/jpa/query-methods.html

AlexeyAntipin commented on src/main/java/faang/school/postservice/repository/PostRepository.java:
```diff
@@ -21,4 +22,6 @@ public interface PostRepository extends CrudRepository<Post, Long> {
     List<Post> findReadyToPublish();
 
-}
+    @Query("SELECT p FROM Post p WHERE p.id IN :ids ORDER BY p.createdAt DESC")
```
Тут тоже без query можно

AlexeyAntipin commented on src/main/java/faang/school/postservice/controller/HashtagController.java:
```diff
@@ -0,0 +1,36 @@
+            @RequestParam(defaultValue = "0") @Min(0) int page,
+            @RequestParam(defaultValue = "20") @Min(1) @Max(100) int size
+    ) {
+        Pageable pageable = PageRequest.of(page, size);
```
Я бы предложил перенести создание Pageable в бизнес логику, потому что это класс именно из Spring Data, к контроллеру с точки зрения чистой архитектуры он не относится

Commit:
- Удалить избыточные комментарии в HashtagServiceImpl
- Заменить List.of() на Collections.emptyList()
- Использовать derived query methods вместо @Query
- Переместить создание Pageable из контроллера в сервис
- Обновить тесты и добавить JavaDoc"
